Documentation about tracing for Knative Eventing. (#1663)
knative/docs: docs/eventing/accessing-traces.md

---
title: "Accessing CloudEvent Traces"
#linkTitle: "OPTIONAL_ALTERNATE_NAV_TITLE"
weight: 15
type: "docs"
---

Depending on the request tracing tool that you have installed on your Knative
Eventing cluster, see the corresponding section for details about how to
visualize and trace your requests.

## Installation

Knative Eventing uses the same tracing plugin as Knative Serving. See the 
[Tracing installation instructions](./../serving/installing-logging-metrics-traces.md#end-to-end-request-tracing)
in the Knative Serving section for details. Note that you do not need to install the 
Knative Serving component itself. To enable request tracing in Knative Eventing 
you need to install only Elasticsearch and either the Zipkin or Jaeger plugins.

## Configuration

Most Knative Eventing tracing configuration is handled by the `config-tracing` ConfigMap in the `knative-eventing` namespace. This single ConfigMap controls tracing for:
 - Brokers
 - Triggers
 - Channel Dispatchers

>      - Channel Dispatchers are the name we use to describe the data plane portion of a Channel. When an event is sent to a Channel, the code handling that event is the Channel's dispatcher. Similarly, when an event is sent from a Channel to one of its Subscribers, the code handling sending that event is the Channel's dispatcher.

Reviewer: Lets separate this description from the list of "tracing controls".
```suggestion

Note: *Channel Dispatcher* describes the data plane portion of a [Channel](./channels/README.md). When an event is sent to a Channel, the code handling that event is the Channel's dispatcher. Similarly, when an event is sent from a Channel to one of its Subscribers, the code handling sending that event is the Channel's dispatcher.
```

Here is an example ConfigMap that samples 10% of all CloudEvents.
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-tracing
  namespace: knative-eventing
data:
  enable: "true"
  zipkin-endpoint: "http://zipkin.istio-system.svc.cluster.local:9411/api/v2/spans"

>   sample-rate: "0.1

Reviewer 2: There is a missing " after `0.1`.

Reviewer 3: @adamharwayne

Reviewer: ```suggestion
  sample-rate: "0.1"
```

```

To see your current configuration:

```shell script
kubectl -n knative-eventing get configmap config-tracing -oyaml
```

Configuration options:

  * `backend`: Valid values are `zipkin`, `stackdriver`, or `none`. The default is `none`.

 * `zipkin-endpoint`: Specifies the URL to the zipkin collector where you want to send the traces. Must be set if backend is set to `zipkin`.

 * `stackdriver-project-id`: Specifies the GCP project into which stackdriver traces will be written. Only has an effect if the back is set to `stackdriver`. If unspecified, the project-id is read from GCP metadata when running on GCP.

 * `sample-rate`: Valid values are one digit decimals from `0` to `1`. Specifies the percentage amount of all the requests that you want to trace.

 * `debug`: Valid values are `true` or `false`. If not specified, defaults to `false`. Set to `true` to enable debug mode, which bypasses sampling and sends all spans to the server.

> Updating the ConfigMap will cause the new configuration to go live almost immediately:

Reviewer: ```suggestion
To edit and then immediately deploy changes to your ConfigMap, run the following command:
```

```shell script
kubectl -n knative-eventing edit configmap config-tracing
```

### Importer Configuration

Most importers do _not_ use the configuration from the shared ConfigMap. Instead, they have a static, 1% sampling rate.

## Tools

To access the traces, you will use a tool such as Zipkin or Jaeger. Follow the Knative Serving [instructions](./../serving/accessing-traces.md) for accessing the traces:
 - [Zipkin](./../serving/accessing-traces.md#zipkin)
 - [Jaeger](./../serving/accessing-traces.md#jaeger)

### Example

For this example, everything happens in the `default` namespace.

> - Broker named `default`

Reviewer: ```suggestion

This example uses the following objects and configuration:

- Broker named `default`.
```

- Two Triggers associated with the Broker:
    - `mutator-t` - Filters to only allow events whose type is `original`. It will reply with an identical event, except the replied event's type will be `mutated`.
    - `display-mutated` - Filters to only allow events whose type is `mutated`.
- An event is sent in with the type `original`.

> So we expect the event to do the  following:

Reviewer: Given the above example objects and configuration, the expected path and behavior of an event is as follows:

---
title: "Accessing CloudEvent Traces"
#linkTitle: "OPTIONAL_ALTERNATE_NAV_TITLE"
weight: 15
type: "docs"
---

Depending on the request tracing tool that you have installed on your Knative
Eventing cluster, see the corresponding section for details about how to
visualize and trace your requests.

> ## Installation

Reviewer: Since the doc  /serving/installing-logging-metrics-traces.md is under the Serving branch, can we clarify by adding an example to that its clear that you dont need to install Serving too? Long term we should move the observability content into its own branch in the website.

Maybe something like:

For example if you want to install Prometheus and Grafana, you can run:
`kubectl apply --filename https://github.com/knative/serving/releases/download/{{< version >}}/monitoring-metrics-prometheus.yaml`

Author: How about something like:

Knative Eventing uses the same tracing installation as Knative Service. Follow Knative Serving's [installation instructions](./../serving/installing-logging-metrics-traces.md) to set up the necessary components. Note that you do not need to install Knative Serving itself. Only Zipkin or Jaeger and its dependencies (e.g. Elasticsearch).

Reviewer: How about the edit i added below?

Reviewer: (1) Also, would it help if the specific install files (.yamls) were listed under Eventing in the "custom install" file too: https://github.com/knative/docs/blob/master/docs/install/Knative-custom-install.md#choosing-knative-installation-files

(2) I guess whats unclear is which .yaml file do you install to get Elasticsearch? For example, for Serving we say to install the ELK stack (monitoring-logs-elasticsearch.yaml) to enable Logging. 
(2a) Is that the file that should be installed with Eventing? 
(2b) And do you get the same Logging features in Eventing?

Author: (1) All of the files that would be listed are already in  "Serving Component and Observability Plugins". Maybe it is worthwhile to split those into two distinct heading, 'Serving Component' and 'Observability Plugins'?

(2a) Yes, you install monitoring-logs-elasticsearch.yaml and either monitoring-tracing-zipkin.yaml or monitoring-tracing-jaeger.yaml (I think the in-mem versions might work without elasticsearch, but I haven't tried).
(2b) I'm not sure, I've never tried them.

Reviewer: I think i will just move all the observability plugins into their own section for now.

Reviewer: PR for separate Observability section: https://github.com/knative/docs/pull/1718

> Knative Eventing uses the same tracing installation as Knative Service. Follow Knative Serving's [installation instructions](./../serving/installing-logging-metrics-traces.md) to set up the necessary components.

Reviewer: ```suggestion
Knative Eventing uses the same tracing plugin as Knative Serving. See the 
[Tracing installation instructions](./../serving/installing-logging-metrics-traces.md#end-to-end-request-tracing)
in the Knative Serving section for details. Note that you do not need to install the 
Knative Serving component itself. To enable request tracing in Knative Eventing 
you need to install only Elasticsearch and either the Zipkin or Jaeger plugins.
```

---
title: "Accessing CloudEvent Traces"
#linkTitle: "OPTIONAL_ALTERNATE_NAV_TITLE"
weight: 15
type: "docs"
---

Depending on the request tracing tool that you have installed on your Knative
Eventing cluster, see the corresponding section for details about how to
visualize and trace your requests.

## Installation

Knative Eventing uses the same tracing plugin as Knative Serving. See the 
[Tracing installation instructions](./../serving/installing-logging-metrics-traces.md#end-to-end-request-tracing)
in the Knative Serving section for details. Note that you do not need to install the 
Knative Serving component itself. To enable request tracing in Knative Eventing 
you need to install only Elasticsearch and either the Zipkin or Jaeger plugins.

## Configuration

Most Knative Eventing tracing configuration is handled by the `config-tracing` ConfigMap in the `knative-eventing` namespace. This single ConfigMap controls tracing for:
 - Brokers
 - Triggers

>  - Channel Dispatchers

Author: I guess we haven't documented it. Logically Channels are broken into two components:
1. Controller - The Control plane work. E.g. modifying the status of K8s object, creating any other needed resources (such as Topics and Subscriptions in GCP PubSub).
1. Dispatecher - The Data plane work. E.g. events sent to the channel go here. Events are sent by the Channel from this piece to the Subscribers.

I hesitate adding it to the overview because it is an implementation detail that can be ignored until you want to debug or trace. So I prefer the overview to be based on the conceptual ideas of Channels, rather than their internal pieces. 

What if I just add an explanation inline here?

Here is an example ConfigMap that samples 10% of all CloudEvents.
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-tracing
  namespace: knative-eventing
data:
  enable: "true"
  zipkin-endpoint: "http://zipkin.istio-system.svc.cluster.local:9411/api/v2/spans"
  sample-rate: "0.1
```

To see your current configuration:

```shell script
kubectl -n knative-eventing get configmap config-tracing -oyaml
```

> The `_example` section explains all the available options.

Author: Things have progressed in the last week. Tweaked your suggestion with the new options.

> Updating the ConfigMap will cause the new configuration to go live almost immediately.

Author: Added the kubectl edit command.

### Importer Configuration

Most importers do _not_ use the configuration from the shared ConfigMap. Instead, they have a static, 1% sampling rate.

## Tools

To access the traces, you will use a tool such as Zipkin or Jaeger. Follow the Knative Serving [instructions](./../serving/accessing-traces.md) for accessing the traces:
 - [Zipkin](./../serving/accessing-traces.md#zipkin)
 - [Jaeger](./../serving/accessing-traces.md#jaeger)

### Example

> Everything happens in the `default` namespace.

Author: Yes. Added.
